File: Plugins/AWS/www/AWS.js
/**
* A consistent interface pattern is applied here.
* 1. Each native method can return success or error.
* 2. The errors are output to the console.log here
* 3. Methods that return data when there is a success
* will return a null when there is an error.
* 4. Native methods that return no data when there is a success
* will return true here and false if there was an error.
*/
"use strict";
var exec = require('cordova/exec');

exports.echo1 = function(message, callback) {
	console.log("AWS.echo1 " + message);
	callback(message);	
};

exports.echo2 = function(message, callback) {
	console.log("INSIDE echo2");
	exec(callback, function(error) {
		AWS.logError("echo2", error, message, null, null);
		callback(error);
	}, "AWS", "echo2", [message]);	
};

exports.echo3 = function(message, callback) {
	console.log("INSIDE echo3");
	exec(callback, function(error) {
		AWS.logError("echo3", error, message, null, null);
		callback(error);
	}, "AWS", "echo3", [message]);	
};

exports.initialize = function(regionName, callback) {
	exec(function() {
		callback(true);
	}, function(error) {
		console.log("ERROR: AWS.initialize " + error);
		callback(false);
	}, "AWS", "initialize", [regionName]);
};

exports.preSignedUrlGET = function(s3Bucket, s3Key, expires, callback) {
    exec(callback, function(error) {
	    AWS.logError("preSignedUrlGET", error, s3Bucket, s3Key, null);
	    callback(null);
    }, "AWS", "preSignedUrlGET", [s3Bucket, s3Key, expires]);
};

exports.preSignedUrlPUT = function(s3Bucket, s3Key, expires, contentType, callback) {
    exec(callback, function(error) {
	    AWS.logError("preSignedUrlPUT", error, s3Bucket, s3Key, null);
	    callback(null);		    
    }, "AWS", "preSignedUrlPUT", [s3Bucket, s3Key, expires, contentType]);
};

exports.downloadText = function(s3Bucket, s3Key, callback) {
    exec(callback, function(error) {
		AWS.logError("downloadText", error, s3Bucket, s3Key, null);
	    callback(null);			    
    }, "AWS", "downloadText", [s3Bucket, s3Key]);
};

exports.downloadData = function(s3Bucket, s3Key, callback) {
    exec(callback, function(error) {
	    AWS.logError("downloadData", error, s3Bucket, s3Key, null);
	    callback(null);			    
    }, "AWS", "downloadData", [s3Bucket, s3Key]);
};

exports.downloadFile = function(s3Bucket, s3Key, filePath, callback) {
    exec(function() {
	    callback(true);
	}, function(error) {
		AWS.logError("downloadFile", error, s3Bucket, s3Key, filePath);
	    callback(false);			    
    }, "AWS", "downloadFile", [s3Bucket, s3Key, filePath]);
};

exports.downloadZipFile = function(s3Bucket, s3Key, filePath, callback) {
    exec(function() {
	    callback(true);
	},  function(error) {
		AWS.logError("downloadZipFile", error, s3Bucket, s3Key, filePath);
	    callback(false);			    
    }, "AWS", "downloadZipFile", [s3Bucket, s3Key, filePath]);
};

exports.uploadVideoAnalytics = function(sessionId, timestamp, data, callback) {
    exec(function() {
	    callback(true);
	}, function(error) {
		AWS.logError("uploadVideoAnalytics", error, sessionId, timestamp, null);
	    callback(false);				
	}, "AWS", "uploadVideoAnalytics", [sessionId, timestamp, data]);
};

exports.uploadText = function(s3Bucket, s3Key, data, callback) {
    exec(function() {
	    callback(true);
	}, function(error) {
		AWS.logError("uploadText", error, s3Bucket, s3Key, null);
	    callback(false);		
	}, "AWS", "uploadText", [s3Bucket, s3Key, data]);
};

exports.uploadData = function(s3Bucket, s3Key, data, contentType, callback) {
    exec(function() {
	    callback(true);
	}, function(error) {
		AWS.logError("uploadData", error, s3Bucket, s3Key, null);
	    callback(false);			
	}, "AWS", "uploadData", [s3Bucket, s3Key, data, contentType]);
};

exports.uploadFile = function(s3Bucket, s3Key, filePath, contentType, callback) {
    exec(function() {
	    callback(true);
    }, function(error) {
	    AWS.logError("uploadFile", error, s3Bucket, s3Key, filePath);
	    callback(false);		    
    }, "AWS", "uploadFile", [s3Bucket, s3Key, filePath, contentType]);
};

exports.zip = function(sourceFile, targetFile, callback) {
    exec(function() { 
	    callback(true); 
	}, function(error) {
	    console.log("ERROR: AWS.zip " + sourceFile + " " + error);
	    callback(false);
    }, "AWS", "zip", [sourceFile, targetFile]);
};

exports.unzip = function(sourceFile, targetDir, callback) {
    exec(function() {
	    callback(true);
	}, function(error) {
	    console.log("ERROR: AWS.unzip " + sourceFile + " " + error);
	    callback(false);		    
    }, "AWS", "unzip", [sourceFile, targetDir]);
};

exports.logError = function(method, error, s3Bucket, s3Key, filePath) {
	var msg = ["\nERROR: AWS."];
	msg.push(method);
	if (s3Bucket) msg.push(" " + s3Bucket);
	if (s3Key) msg.push("." + s3Key);
	if (filePath) msg.push(" " + filePath);
	msg.push(" -> " + error);
	console.log(msg.join(""));	
};

File: Plugins/AWS/src/ios/AWS.swift
//
//  AWS.swift
//  AWS
//
//  Created by Gary Griswold on 5/15/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//
/**
* This class is the cordova native interface code that calls the AwsS3.
* It is a thin wrapper around the AwsS3 class that AwsS3 can also 
* be used directly by other .swift classes.
*/
//import AWS used for AWS.framework

@objc(AWS) class AWS : CDVPlugin {
						    
    @objc(initializeRegion:)
    func initializeRegion(command:  CDVInvokedUrlCommand) {
        let regionName = command.arguments[0] as? String ?? ""
        AwsS3.region = regionName
        let result = CDVPluginResult(status: CDVCommandStatus_OK)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }
	
	@objc(echo2:) 
	func echo2(command:  CDVInvokedUrlCommand) {
		let message = command.arguments[0] as? String ?? ""
		let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: message)
		self.commandDelegate!.send(result, callbackId: command.callbackId)
	}
	
	@objc(echo3:)
	func echo3(command: CDVInvokedUrlCommand) {
		let message = command.arguments[0] as? String ?? ""
		let response = AwsS3.shared.echo3(message: message);
		let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: response)
		self.commandDelegate!.send(result, callbackId: command.callbackId)		
	}
	
	@objc(preSignedUrlGET:) 
	func preSignedUrlGET(command: CDVInvokedUrlCommand) {
		AwsS3.shared.preSignedUrlGET(
			s3Bucket: command.arguments[0] as? String ?? "",
			s3Key: command.arguments[1] as? String ?? "",
			expires: command.arguments[2] as? Int ?? 3600,
			complete: { url in 
				let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: url?.absoluteString)
				self.commandDelegate!.send(result, callbackId: command.callbackId)
			}
		)
	}
	
	@objc(preSignedUrlPUT:) 
	func preSignedUrlPUT(command: CDVInvokedUrlCommand) {
		AwsS3.shared.preSignedUrlPUT(
			s3Bucket: command.arguments[0] as? String ?? "",
			s3Key: command.arguments[1] as? String ?? "",
			expires: command.arguments[2] as? Int ?? 3600,
			contentType: command.arguments[3] as? String ?? "",
			complete: { url in 
				let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: url?.absoluteString)
				self.commandDelegate!.send(result, callbackId: command.callbackId)
			}
		)
	}

	@objc(downloadText:) 
	func downloadText(command: CDVInvokedUrlCommand) {
		AwsS3.shared.downloadText(
			s3Bucket: command.arguments[0] as? String ?? "",
			s3Key: command.arguments[1] as? String ?? "",
            complete: { error, data in
            	var result: CDVPluginResult
            	if let err = error {
	            	result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: err.localizedDescription)
            	} else {
	            	result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: data)
            	}
            	self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        )
    }
    
    @objc(downloadData:) 
    func downloadData(command: CDVInvokedUrlCommand) {	
		AwsS3.shared.downloadData(
			s3Bucket: command.arguments[0] as? String ?? "",
			s3Key: command.arguments[1] as? String ?? "",
            complete: { error, data in
	            var result: CDVPluginResult
	            if let err = error {
		            result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: err.localizedDescription)
	            } else {
		            result = CDVPluginResult(status: CDVCommandStatus_OK, messageAsArrayBuffer: data)
	            }
	            self.commandDelegate!.send(result, callbackId: command.callbackId)
	        }
        )
    }
    
    @objc(downloadFile:) 
    func downloadFile(command: CDVInvokedUrlCommand) {
	    print("Documents \(NSHomeDirectory())") 
	    let filePath: String = command.arguments[2] as? String ?? ""
	    AwsS3.shared.downloadFile(
			s3Bucket: command.arguments[0] as? String ?? "",
			s3Key: command.arguments[1] as? String ?? "",
			filePath: URL(fileURLWithPath: NSHomeDirectory() + filePath),
            complete: { error in
	            var result: CDVPluginResult
		        if let err = error {
		            result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: err.localizedDescription)
	            } else {
		            result = CDVPluginResult(status: CDVCommandStatus_OK)
	            }
	            self.commandDelegate!.send(result, callbackId: command.callbackId)
	        }	    
	    )
    }
    @objc(downloadZipFile:) 
    func downloadZipFile(command: CDVInvokedUrlCommand) {
	    print("Documents \(NSHomeDirectory())") 
	    let filePath: String = command.arguments[2] as? String ?? ""
	    AwsS3.shared.downloadZipFile(
			s3Bucket: command.arguments[0] as? String ?? "",
			s3Key: command.arguments[1] as? String ?? "",
			filePath: URL(fileURLWithPath: NSHomeDirectory() + filePath),
			view: nil,
            complete: { error in
	            var result: CDVPluginResult
	            if let err = error {
		            result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: err.localizedDescription)
	            } else {
		            result = CDVPluginResult(status: CDVCommandStatus_OK)
	            }
	            self.commandDelegate!.send(result, callbackId: command.callbackId)
	        }	    
	    )
    }
    @objc(uploadAnalytics:) 
    func uploadVideoAnalytics(command: CDVInvokedUrlCommand) {
	    let data = command.arguments[3] as? String ?? ""
	    AwsS3.shared.uploadAnalytics(
		    sessionId: command.arguments[0] as? String ?? "", 
		    timestamp: command.arguments[1] as? String ?? "",
		    prefix: command.arguments[2] as? String ?? "",
		    json: data.data(using: String.Encoding.utf8)!,
            complete: { error in
	            var result: CDVPluginResult
	            if let err = error {
		            result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: err.localizedDescription)
	            } else {
		            result = CDVPluginResult(status: CDVCommandStatus_OK)
	            }
	            self.commandDelegate!.send(result, callbackId: command.callbackId)
	        }		    
	    )
    }
    
    @objc(uploadText:) 
    func uploadText(command: CDVInvokedUrlCommand) {    
	    AwsS3.shared.uploadText(
			s3Bucket: command.arguments[0] as? String ?? "",
			s3Key: command.arguments[1] as? String ?? "",
		    data: command.arguments[2] as? String ?? "",
		    contentType: command.arguments[3] as? String ?? "",
            complete: { error in
	            var result: CDVPluginResult
	            if let err = error {
		            result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: err.localizedDescription)
	            } else {
		            result = CDVPluginResult(status: CDVCommandStatus_OK)
	            }
	            self.commandDelegate!.send(result, callbackId: command.callbackId)
	        }	    
	    )
    }
    
    @objc(uploadData:) 
    func uploadData(command: CDVInvokedUrlCommand) {
	    AwsS3.shared.uploadData(
			s3Bucket: command.arguments[0] as? String ?? "",
			s3Key: command.arguments[1] as? String ?? "",
		    data: command.arguments[2] as? Data ?? Data(),
		    contentType: command.arguments[3] as? String ?? "",
            complete: { error in
	            var result: CDVPluginResult
	            if let err = error {
		            result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: err.localizedDescription)
	            } else {
		            result = CDVPluginResult(status: CDVCommandStatus_OK)
	            }
	            self.commandDelegate!.send(result, callbackId: command.callbackId)
	        }		    
	    )
    }
    
    /**
    * Warning: this does not use the uploadFile method of TransferUtility,
	* See note in AwsS3.uploadFile for more info.
    */
    @objc(uploadFile:) 
    func uploadFile(command: CDVInvokedUrlCommand) {
	    let filePath = command.arguments[2] as? String ?? ""
	    AwsS3.shared.uploadFile(
			s3Bucket: command.arguments[0] as? String ?? "",
			s3Key: command.arguments[1] as? String ?? "",
			filePath: URL(fileURLWithPath: NSHomeDirectory() + filePath),
			contentType: command.arguments[3] as? String ?? "",
            complete: { error in
            	var result: CDVPluginResult
				if let err = error {
	            	result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: err.localizedDescription)
				} else {
	            	result = CDVPluginResult(status: CDVCommandStatus_OK)
				}
				self.commandDelegate!.send(result, callbackId: command.callbackId)
			}
	    )
    }
}		
